Add syncplay feature

OancaAndrei opened the pull request:
**Changes**
This PR adds playback syncing of different sessions that are watching the same content.
- Users that are playing something can create a group
- Multiple users can join a group to sync the same media
- Playback (play, pause, seek) is synced on all players of a group

The implementation tries to be time and network resistant, meaning that syncing is not based on merely sending the _playback command_ but also tells the clients _when_ to actually run said command. This means shifting the network latency problem to a time synchronization one.

Ideally, the client's time should be correct but, in the event this isn't true, a server time estimation is made by the client (clients always use this at the moment).

A major problem faced was the delay caused by issuing a play command on the web-player and it actually starting the playback. This issue has been solved by speeding up the playback to catch up.
In other words, to implement this feature on other players, the "playback rate" (or speed) should be supported by the player to achieve optimal results.

Feedback and change proposals are welcome.
I'm curious to see whether the current UX is ideal or if some things should be revised.

**Planned features for this PR**
- [x] Option to toggle feature for specific user [User edit > Media Playback > Syncplay access]
- [x] Filter groups by library access and parental rating (needs review)

**Proposed features for future PRs**
- Share the transcoding process with the members of the group
- Manually adjust delay
- Time sync between devices in the same network
- Group types (public, private, personal)
- Group join by PIN

**Features being tackled in next PR**
- Sync playlist as well (more discussion needed)
- Group wait when one or more members of a group are buffering

Please check [this](https://github.com/jellyfin/jellyfin-web/pull/1011) related PR as well.

**Issues**
Addresses some requested features of #227.

**Docker build**
```
docker pull andreioanca/jellyfin:syncplay
```

PrplHaz4 commented:
Take this with a grain of salt - I'm not an authority on any of this...

Did you look into creating this as a plugin?  Guessing some of the UI stuff might make that tough to do?

Also, HasAccessToItem seems like it duplicates some of the code here: https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.Api/UserLibrary/ItemsService.cs#L182

IMO, Item access code should be centralized and reusable reducing the likelihood of permissions problems.  A method like HasAccesToItem shouldn't be scoped (I don't think?) to Syncplay only...

sparky8251 commented:
This is fine to be in core. Not just because it needs to manipulate the frontend and there is no plugin system there, but because this is a feature that falls in the scope of Jellyfin as a media server.

OancaAndrei commented:
> Also, HasAccessToItem seems like it duplicates some of the code here: https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.Api/UserLibrary/ItemsService.cs#L182
> 
> IMO, Item access code should be centralized and reusable reducing the likelihood of permissions problems. A method like HasAccesToItem shouldn't be scoped (I don't think?) to Syncplay only...

Indeed, I agree that such a method should be placed someplace else, available to other modules as well.
I think that `HasAccesToItem` will eventually end up checking several things, like library access, parental rating, access schedule and related.

gnouts commented:
Hi,
Out of curisioty, I tried to test your PR.
I build jellyfin using the Dockerfile, replacing jellyfin-web with your related PR.
I got a jellyfin with Sync enable for my users in their profile. But I can only see the new icon (top right corner) that say "No groups available".
Although I'm not developer, I read your commits and I guess I should have a new entry in the movie detail (somewhere near "Download" and "Copy stream url") ?
What am I missing ? :)

JustAMan commented:
@OancaAndrei can you clarify what's left to mark this as ready to review?

OancaAndrei commented:
> @OancaAndrei can you clarify what's left to mark this as ready to review?

@JustAMan Sure, to avoid duplication I moved the progress tracking to the other PR, but I can copy it here as well, if it makes things more clear.

At the moment I'm investigating some issues that appear on Safari browsers.
On a more backend related view, what's currently planned are also public/private groups and PIN protected joining.

JustAMan commented:
> On a more backend related view, what's currently planned are also public/private groups and PIN protected joining.

Those can come as separate PRs enhancing the feature I think. Let's focus on bringing it in the code in the simplest possible form (which allows extending) and improve it later.

gurupras commented:
Sorry, I'm very new to Jellyfin, but would all participants need to have an account on the Jellyfin server, and need to be logged in?

JustAMan commented:
@OancaAndrei I just gave a look at your "todo list" for this PR, and I have a question - can we mark what we have now as "minimal usable" and start merging, and other boxes will be ticked by further PRs? I need to remind that it's not only a burden on you to implement stuff but also a burden on us to review it, so it's waaaaaaaaaaay better to have a thing come in several atomic PRs (each of them should ideally leave the product functional, even if not at the completely desired state) than in one giant PR that would be reviewed for weeks.

I must ask again, what is left on the server part to mark it as a first usable stage? IMHO the first usable stage is when syncplay is possible to be turned on/off globally by server admin and is joinable by anyone on the server. No groups types, no group pins, not even waiting is exactly needed. People can circumvent waiting by pausing manually (they're likely to be over some other communication channel to be able to ask to pause or pause themselves if it buffers).

So we can get this PR in, and _then_ build PRs that improve and polish the feature.

OancaAndrei commented:
> I need to remind that it's not only a burden on you to implement stuff but also a burden on us to review it, so it's waaaaaaaaaaay better to have a thing come in several atomic PRs (each of them should ideally leave the product functional, even if not at the completely desired state) than in one giant PR that would be reviewed for weeks.

I totally agree. I'm pretty happy with the current status of the PR so I think it's ready for review. I didn't mean to overload you with any excessive work.

[This](https://github.com/jellyfin/jellyfin-web/pull/1011#issuecomment-613141147) comment lead me to think that the new features had to be well defined before any merging occurred. And, because the required features by the UI implied changes to this PR as well, I was looking for a structure (group types / participant types) general enough to fit the different usages.

This PR also includes another PR regarding `KeepAlive` of WebSockets. Only now I realize that I should have mentioned it here as well, [link](https://github.com/jellyfin/jellyfin-web/pull/1067#issuecomment-615419028) to other PR.
Something that's missing is how WebSockets are handled server-side when "marked" as lost, although I don't consider it to be a blocking thing for this PR.

> IMHO the first usable stage is when syncplay is possible to be turned on/off globally by server admin and is joinable by anyone on the server.

At the moment, Syncplay can be turned off selectively for single users, while joining a group is restricted to library access and parental rating.

If you think that it's ready for review then I'll mark it as such :grin:.

Also, should the [related](https://github.com/jellyfin/jellyfin-web/pull/1011) PR be marked as _ready for review_ too?

JustAMan commented:
@OancaAndrei I'm officially declaring this as ready! :) @jellyfin/backend let's review and get this awesome feature in :)

JustAMan commented:
> Also, should the [related](https://github.com/jellyfin/jellyfin-web/pull/1011) PR be marked as _ready for review_ too?

That depends on if its state corresponds to the state of this PR, like, if it fits the criteria I have listed [above](https://github.com/jellyfin/jellyfin/pull/2733#issuecomment-616207511).

OancaAndrei commented:
> That depends on if its state corresponds to the state of this PR, like, if it fits the criteria I have listed [above](https://github.com/jellyfin/jellyfin/pull/2733#issuecomment-616207511).

It does, I tried to keep both PRs synced (haha). My only concern was that point about the UI changing in the future.

JustAMan commented on MediaBrowser.Api/Syncplay/SyncplayService.cs:
```diff
@@ -0,0 +1,267 @@
+        public string SessionId { get; set; }
+
+        [ApiMember(Name = "When", IsRequired = true, DataType = "string", ParameterType = "query", Verb = "POST")]
+        public string When { get; set; }
```
`When` and `Resume` could use a help comment IMHO

JustAMan commented on MediaBrowser.Api/Syncplay/SyncplayService.cs:
```diff
@@ -0,0 +1,267 @@
+            }
+            catch (ArgumentNullException)
+            {
+                // Do nothing
```
can we at least log warnings or something here?

OancaAndrei replied:
The "specification" I had in mind was that both `null` and empty strings would be considered by the server to be valid. It is fine for a client to request to join a group without sending a value for `PlayingItemId`: it means that the client isn't playing anything.

Should I pick an option and consider the other as invalid and thus log it?

JustAMan replied:
This (what you wrote in an answer) is an excellent comment, much better than `// Do nothing` in the code currently :)
Please replace this `// Do nothing` with the real explanation of why we're ignoring the exception.

Or maybe even better check explicitly, like
```cs
if (!String.IsNullOrEmpty(request.PlayingItemId))
{
    try
    {
        joinRequest.PlayingItemId = Guid.Parse(request.PlayingItemId);
    }
    catch (... whatever ...)
    {
        // log error, quit method - this is clearly a malformed request
    }
}

JustAMan commented on MediaBrowser.Api/Syncplay/TimeSyncService.cs:
```diff
@@ -0,0 +1,68 @@
+namespace MediaBrowser.Api.Syncplay
+{
+    [Route("/GetUtcTime", "GET", Summary = "Get UtcTime")]
+    [Authenticated]
```
why auth is required?

OancaAndrei replied:
No reason in particular. I just thought that having a session might come handy in the future.
Do you think that it should be freely available instead?

dkanada replied:
I think we should put as few endpoints in front of authentication as possible. We currently have a few and I still think that's too many. :laughing:

JustAMan commented on MediaBrowser.Api/Syncplay/TimeSyncService.cs:
```diff
@@ -0,0 +1,68 @@
+
+            var response = new UtcTimeResponse();
+            response.RequestReceptionTime = requestReceptionTime;
+            var currentSession = GetSession(_sessionContext);
```
`currentSession` seems unused?

OancaAndrei replied:
I forgot to remove it. I was testing some "computation load" between the request being received and the response being sent.

JustAMan commented on MediaBrowser.Controller/Syncplay/GroupInfo.cs:
```diff
@@ -0,0 +1,148 @@
+        /// <value>The last activity.</value>
+        public DateTime LastActivity { get; set; }
+
+        /// <summary>
+        /// Gets the partecipants.
+        /// </summary>
+        /// <value>The partecipants.</value>
+        public readonly ConcurrentDictionary<string, GroupMember> Partecipants =
+        new ConcurrentDictionary<string, GroupMember>(StringComparer.OrdinalIgnoreCase);
```
```suggestion
        /// <summary>
        /// Gets the participants.
        /// </summary>
        /// <value>The participants, or members of the group.</value>
        public readonly ConcurrentDictionary<string, GroupMember> Participants =
            new ConcurrentDictionary<string, GroupMember>(StringComparer.OrdinalIgnoreCase);
```

OancaAndrei replied:
All these years... :laughing:

JustAMan commented on MediaBrowser.Model/Syncplay/GroupInfoModel.cs:
```diff
@@ -0,0 +1,38 @@
+        /// Gets or sets the partecipants.
+        /// </summary>
+        /// <value>The partecipants.</value>
+        public string[] Partecipants { get; set; }
```
```suggestion
        public string[] Participants { get; set; }
```

JustAMan commented on MediaBrowser.Model/Syncplay/SendCommandType.cs:
```diff
@@ -0,0 +1,21 @@
+        /// <summary>
+        /// The seek command. Instructs users to seek to a specified time.
+        /// </summary>
+        Seek = 2
```
how do I request a "next item" or just "different item than is currently played"?

OancaAndrei replied:
There's no way at the moment. Changing the current item means creating a new group. In the future this will be handled by the shared playlist I guess.

OancaAndrei commented on Emby.Server.Implementations/HttpServer/WebSocketConnection.cs:
```diff
@@ -174,7 +172,15 @@ private void OnReceiveInternal(string message)
+                if (info.MessageType.Equals("KeepAlive", StringComparison.Ordinal))
+                {
+                    SendKeepAliveResponse();
+                }
+
+                if (OnReceive != null)
+                {
+                    OnReceive(info);
+                }
```
The idea was to make it visible to higher levels, but I guess they don't really care

OancaAndrei commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +130,130 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+        /// </summary>
+        /// <param name="sender">The WebSocket.</param>
+        /// <param name="e">The event arguments.</param>
+        private void _webSocket_Closed(object sender, EventArgs e)
```
I see, I was following the style of `_serverManager_WebSocketConnected` that's a few lines above, but I like it better this way (`OnWebSocketClosed`).

Does `_serverManager_WebSocketConnected` follow another style because of `_serverManager` being a singleton? Or should all event handlers follow `OnEvent`? If the latter then I'll update the others as well (those added by the PR).

OancaAndrei replied:
Nevermind, I have now seen the proposed changes bellow for `_serverManager_*` as well.

JustAMan replied:
I don't see you renaming `_serverManager_WebSocketConnected`. Not that it was required, but would be nice to use consistent style :)

OancaAndrei replied:
Oh ok, I've renamed those in `SyncaplayManager`, renaming `_serverManager_WebSocketConnected` didn't even cross my mind as I thought it isn't related to the PR.

OancaAndrei commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +130,130 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+            }
+            catch (WebSocketException exception)
+            {
+                _logger.LogDebug(exception, "Error sending ForceKeepAlive message to WebSocket.");
```
I see an error as something that originates from inside the server itself.
In this case, I would consider a network failure not that relevant to the server as a whole. I mean, it's not rare for a connection to drop (be it for network issues or because the browser has been closed), thereby the server should handle it gracefully but not complain that much.

Also, `ForceKeepAlive` is sent when the client hasn't responded in a while, so the server is already expecting a possible failure.

I would have opted for a warning if it wasn't for the fact that the dropped connection will be notified to the interested session that could potentially become offline. So the idea was to avoid two messages caused by the same event.

Don't know if this makes sense :sweat_smile:. These were my considerations, but if you think that being more verbose is better then I would agree as well.

OancaAndrei replied:
Nevermind, I've mistaken that line for the one that comes after. In this case it does make sense to print an error as the connection has just been established.

OancaAndrei commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +130,130 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+        /// </summary>
+        private void StartKeepAliveTimer()
+        {
+            if (_keepAliveTimer == null)
```
Nice catch, I should probably revise all calls as I haven't really taken into consideration the async nature of requests.

OancaAndrei commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -0,0 +1,417 @@
+        /// <inheritdoc />
+        public void SessionJoin(SessionInfo session, JoinGroupRequest request)
+        {
+            if (session.NowPlayingItem != null &&
+                session.NowPlayingItem.Id.Equals(_group.PlayingItem.Id) &&
```
Definitely hahah, as you might have probably noticed, I have never used C# before

JustAMan replied:
My main languages (besides human) are C and Python, I'm learning as I go :) This project is the first big C# thing I do, so we're in the same boat, I've just jumped in earlier 😃

OancaAndrei commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -0,0 +1,417 @@
+            else
+            {
+                var playRequest = new PlayRequest();
+                playRequest.ItemIds = new Guid[] { _group.PlayingItem.Id };
```
Exactly, only the current item that's playing in one's session.
Should I create a custom "response" class for this? `PlayRequest` seemed the right choice, considering future expansion.

JustAMan replied:
Not in this PR then. We can tweak it later.

OancaAndrei commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -0,0 +1,417 @@
+                var command = NewSyncplayCommand(SendCommandType.Seek);
+                SendCommand(session, BroadcastType.AllGroup, command);
+            }
+            // TODO: client does not implement this yet
```
It was more of a note to myself signaling that this part is probably going to change

OancaAndrei commented on Emby.Server.Implementations/Syncplay/SyncplayManager.cs:
```diff
@@ -0,0 +1,331 @@
+
+        void _sessionManager_PlaybackStopped(object sender, PlaybackStopEventArgs e)
+        {
+            var session = e.Session;
+            if (!IsSessionInGroup(session)) return;
+            LeaveGroup(session);
```
At the moment it is. It will surely change once playlist sharing is added.

OancaAndrei commented on Emby.Server.Implementations/Syncplay/SyncplayManager.cs:
```diff
@@ -0,0 +1,331 @@
+                    folder => folder.Id.ToString("N", CultureInfo.InvariantCulture)
+                );
+                var intersect = collections.Intersect(user.Policy.EnabledFolders);
+                return intersect.Count() > 0 && hasParentalRatingAccess;
```
Definitely an improvement, strange I missed that :sweat_smile:

OancaAndrei commented on Emby.Server.Implementations/Syncplay/SyncplayManager.cs:
```diff
@@ -0,0 +1,331 @@
+            if (session.NowPlayingItem != null)
+            {
+                return _groups.Values.Where(
+                    group => HasAccessToItem(user, group.GetPlayingItemId())
+                ).Where(
+                    group => group.GetPlayingItemId().Equals(session.FullNowPlayingItem.Id)
```
It felt "cleaner" but now I realize that it's not. It's also slower :sweat_smile:

OancaAndrei commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -0,0 +1,417 @@
+                SessionInfo[] sessions = FilterSessions(from, type);
+                foreach (var session in sessions)
+                {
+                    yield return _sessionManager.SendSyncplayGroupUpdate(session.Id.ToString(), message, CancellationToken.None);
```
Honestly, I haven't looked at how `Task`s work. I'll do my research :grin:

OancaAndrei replied:
So basically, I should just add a parameter for the cancellation token? Also, where should this token come from? From the controller itself or from the service?

It's still unclear to me when these tokens are needed.

JustAMan replied:
This place now looks correct, I'll see about the callers of this API later when I read that code (your PR is big mate :D)

OancaAndrei commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +130,130 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+                _logger.LogDebug("Lost {0} WebSockets.", lost.Count());
+            }
+
+            if (!_webSockets.Any())
```
Also, what should happen to the related session?

JustAMan replied:
No idea. Probably it should be closed...

OancaAndrei commented on Emby.Server.Implementations/Syncplay/SyncplayManager.cs:
```diff
@@ -0,0 +1,331 @@
+
+            if (group.IsGroupEmpty())
+            {
+                _groups.Remove(group.GetGroupId().ToString(), out _);
```
I see, I've been indeed relying on the concurrent dictionary doing the hard work. I'll have to investigate.

OancaAndrei replied:
Done.

JustAMan commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -55,6 +94,7 @@ void _serverManager_WebSocketConnected(object sender, GenericEventArgs<IWebSocke
             if (session != null)
             {
                 EnsureController(session, e.Argument);
+                var task = KeepAliveWebSocket(e.Argument);
```
where `task` is used?

OancaAndrei replied:
Nowhere. I had to add it in order to suppress the `Because this call is not awaited, execution of the current method continues before the call is completed` issue on Sonar. `In that case [no await for async call], you can suppress the warning by assigning the task result of the call to a variable.`

Should `KeepAliveWebSocket` be awaited instead?
Actually, now that I'm seeing the other comment bellow regarding the shown warning, I'm considering making `KeepAliveWebSocket` non-async, then letting the rather rare exception returned when trying to send a message be ignored and finally letting the watcher handle the WebSocket that will eventually time out.

JustAMan replied:
I think the better way is really making `KeepAliveWebSocket` non-async.

JustAMan commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +140,205 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+        private void OnWebSocketClosed(object sender, EventArgs e)
+        {
+            var webSocket = (IWebSocketConnection) sender;
+            _logger.LogDebug("WebSockets {0} closed.", webSocket);
```
```suggestion
            _logger.LogDebug("WebSocket {0} is closed.", webSocket);
```

OancaAndrei commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +140,205 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+            }
+            catch (WebSocketException exception)
+            {
+                _logger.LogWarning(exception, "Error sending ForceKeepAlive message to WebSocket {0}.", webSocket);
```
Again, I'm considering making `KeepAliveWebSocket` non-async, then letting the rather rare exception returned when trying to send a message be ignored and finally letting the watcher handle the WebSocket that will eventually time out.

In other words, the try-catch will be gone. Does it sound reasonable?

JustAMan replied:
Well, I don't usually like ignoring exceptions... I would still rather we log it, because it might help in future debugging.

OancaAndrei replied:
I wasn't aware of the existence of `.Wait()`, I thought that only `async / await` were available. This simplifies things :sweat_smile:.

OancaAndrei commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +140,205 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+            {
+                if (_keepAliveCancellationToken == null)
+                {
+                    _keepAliveCancellationToken = new CancellationTokenSource();
```
It's being cancelled in `StopKeepAlive`, which in turn is called when no more sockets are connected or when `Dispose`-ing. Am I missing something?

JustAMan replied:
probably not... I have seen the other parts after I published my previous comment.

JustAMan commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -136,85 +146,121 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+                {
+                    _keepAliveCancellationToken = new CancellationTokenSource();
+                    // Start KeepAlive watcher
+                    var task = RepeatAsyncCallbackEvery(
```
why declare a `task` if you're not using it afterwards? doesn't C# allow you to _not_ assign a return value if you don't need it?

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -228,190 +271,271 @@ public void SessionJoin(SessionInfo session, JoinGroupRequest request)
+            // Sanitize PositionTicks
+            var ticks = SanitizePositionTicks(request.PositionTicks);
+
+            // Pause and seek
+            _group.IsPaused = true;
```
why pause?..

OancaAndrei replied:
Mainly because of network bandwidth, which in turn affects when clients are ready. Seeking and expecting players to be playing the media that fast results in the content being skipped. The time clients spend on buffering is counted by the server as "playing time".

This will probably change when clients implement "group waiting": that functionality hasn't been tested yet so I don't really know how it behaves.

So, should this change be made now, as to not break the "specs" later, or is it fine changing it later?

JustAMan replied:
We can change it later. Spec will only get finalized in a sense only when a release is made. We can always do a few more PRs until then :)

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -228,190 +271,271 @@ public void SessionJoin(SessionInfo session, JoinGroupRequest request)
+                var command = NewSyncplayCommand(SendCommandType.Pause);
+                SendCommand(session, BroadcastType.AllReady, command, CancellationToken.None);
+
+                var updateOthers = NewSyncplayGroupUpdate(GroupUpdateType.GroupWait, session.UserName);
```
what if there are multiple sessions from same user, only one of which is buffering?

OancaAndrei replied:
This is merely a message for clients to know why the media is getting paused. Instead of showing a general "Someone is buffering" I thought of showing who's having problems.

A nice idea, for the future, would be to add a "group monitor tab" that shows the status of connected clients (user, device, buffering status, ping, etc.). In that case a more precise distinction would be needed.

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -228,190 +271,271 @@ public void SessionJoin(SessionInfo session, JoinGroupRequest request)
+
+                if (_group.IsBuffering())
                 {
+                    // Others are buffering, tell this client to pause when ready
```
```suggestion
                    // Others are still buffering, tell this client to pause when ready
```

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -228,190 +271,271 @@ public void SessionJoin(SessionInfo session, JoinGroupRequest request)
-                    // Make sure client has latest group state
-                    var command = NewSyncplayCommand(SendCommandType.Play);
-                    SendCommand(session, BroadcastType.SingleSession, command);
+                        // Client, that was buffering, resumed playback but did not update others in time
```
nice catch!

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -228,190 +271,271 @@ public void SessionJoin(SessionInfo session, JoinGroupRequest request)
+            ticks = ticks >= 0 ? ticks : 0;
+            if (_group.PlayingItem.RunTimeTicks != null)
+            {
+                var runTimeTicks = _group.PlayingItem.RunTimeTicks ??= 0;
+                ticks = ticks > runTimeTicks ? runTimeTicks : ticks;
```
```suggestion
                ticks = ticks > _group.PlayingItem.RunTimeTicks ? _group.PlayingItem.RunTimeTicks : ticks;
```
you've already checked that `_group.PlayingItem.RunTimeTicks` is not `null`, why check again?..

OancaAndrei replied:
The first check is for `null`, the other ensures that `ticks` never go beyond `_group.PlayingItem.RunTimeTicks`. It looked more clear to me :sweat_smile:.

The `??=` needs to be changed though, as you pointed out bellow.

OancaAndrei replied:
My bad, that's indeed checking twice the same thing.

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -228,190 +271,271 @@ public void SessionJoin(SessionInfo session, JoinGroupRequest request)
+        private void HandlePingUpdateRequest(SessionInfo session, PlaybackRequest request)
+        {
+            // Collected pings are used to account for network latency when unpausing playback
+            var ping = request.Ping ??= _group.DefaulPing;
```
Why are you using `??=` instead of `??` everywhere? AFAIK it updates the field if it was `null`, why do we want to update `request` during its handling? I would assume request should be considered read-only...

OancaAndrei replied:
In my quick google-search I misread what `??=` does. `??` is indeed what I want in the majority of the cases, I'll review them all.

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -228,190 +271,271 @@ public void SessionJoin(SessionInfo session, JoinGroupRequest request)
+                PlayingItemName = _group.PlayingItem.Name,
+                PlayingItemId = _group.PlayingItem.Id.ToString(),
+                PositionTicks = _group.PositionTicks,
+                Participants = _group.Participants.Values.Select(session => session.Session.UserName).ToArray()
```
so we identify sessions by their users' names, not by some session ids? why?

OancaAndrei replied:
This is similar to the "buffering message", it's meant (for now) to be seen at a user level, specifically when joining a group, the user sees this list of participants.

I just noticed that there might be duplicates of the same user that could be removed to keep the list cleaner.

In the future, more detailed info could be sent to the clients.

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayManager.cs:
```diff
@@ -165,19 +175,28 @@ public void NewGroup(SessionInfo session)
+                {
+                    Type = GroupUpdateType.CreateGroupDenied
+                };
+                _sessionManager.SendSyncplayGroupUpdate(session.Id.ToString(), error, CancellationToken.None);
```
maybe `NewGroup()` should also accept the cancellation token? I think that ideally the whole chain starting from the root HTTP request handler should pass down the cancellation token...

applicable here and everywhere else being called from public HTTP API

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayManager.cs:
```diff
@@ -187,59 +206,83 @@ public void JoinGroup(SessionInfo session, string groupId, JoinGroupRequest requ
+
+                    var error = new GroupUpdate<string>()
+                    {
+                        Type = GroupUpdateType.GroupNotJoined
```
error name looks misleading, should probably be `GroupDoesNotExist`

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayManager.cs:
```diff
@@ -187,59 +206,83 @@ public void JoinGroup(SessionInfo session, string groupId, JoinGroupRequest requ
+
+                if (group.IsGroupEmpty())
+                {
+                    _groups.Remove(group.GetGroupId().ToString(), out _);
```
I suggest to add a logging entry here about group removal

JustAMan commented on MediaBrowser.Api/Syncplay/SyncplayService.cs:
```diff
@@ -90,12 +90,20 @@ public class SyncplayBufferingRequest : IReturnVoid
+        /// Gets or sets whether this is a buffering or a buffering-done request.
+        /// </summary>
+        /// <value><c>true</c> if buffering is complete; <c>false</c> otherwise.</value>
         [ApiMember(Name = "Resume", IsRequired = true, DataType = "bool", ParameterType = "query", Verb = "POST")]
```
given the doc above I start thinking the field should be named `BufferingDone` or something similar... especially seeing `PlaybackRequestType` has a value named `BufferingDone` as well :D

JustAMan commented on MediaBrowser.Controller/Syncplay/GroupInfo.cs:
```diff
@@ -46,19 +46,19 @@ public class GroupInfo
-        public readonly ConcurrentDictionary<string, GroupMember> Partecipants =
-        new ConcurrentDictionary<string, GroupMember>(StringComparer.OrdinalIgnoreCase);
+        /// <value>The participants, or members of the group.</value>
+        public readonly ConcurrentDictionary<string, GroupMember> Participants =
```
I'm starting to feel wary of data races here as well... example - computing highest ping and in parallel getting a new client joining the group might produce unexpected results.
So it's probably better to do a manual locking here, too.

OancaAndrei replied:
> computing highest ping and in parallel getting a new client joining the group

I don't think that can happen anymore as `SyncplayManager` handles the locking now.
I should probably replace the `ConcurrentDictionary` as it's not really needed.

Shall I add a notice somewhere stating that all operations of `SyncplayController` must happen inside a lock? Or do you consider safer introducing another lock?

JustAMan replied:
It's better to avoid nested locking, because doing it correctly is _hard_. Just add a notice that outside locking is needed.

JustAMan commented on MediaBrowser.Model/Syncplay/GroupUpdateType.cs:
```diff
@@ -1,37 +1,53 @@
         /// The user-joined update. Tells members of a group about a new user.
         /// </summary>
-        UserJoined = 0,
+        UserJoined,
```
why did you remove the values?

OancaAndrei replied:
I found myself adding new events and forgetting to edit these values.
Do those values have a meaning? Like, is it necessary to explicit them instead of letting C# pick one?

JustAMan replied:
No idea tbh... :) cc @jellyfin/backend for more comments

JustAMan commented on MediaBrowser.Model/Syncplay/GroupUpdateType.cs:
```diff
@@ -1,37 +1,53 @@
+        PrepareSession,
         /// <summary>
-        /// The not-in-group update. Tells a user that no group has been joined.
+        /// The not-in-group error. Tells a user that it doesn't belong to a group.
```
```suggestion
        /// The not-in-group error. Tells a user that they don't belong to a group.
```
calling a user "it" doesn't sound polite 🤣

OancaAndrei replied:
How rude of me :laughing:

OancaAndrei commented:
> Please don't consider me a nitpicker, it's just the request is big, and we're gradually polishing it to a perfect state.

Not at all :smile:, that's exactly what a reviewer should do in order to keep the code base clean.

OancaAndrei commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -95,131 +102,167 @@ private void CheckDisposed()
+            SendGroupUpdate(session, BroadcastType.CurrentSession, updateSession, CancellationToken.None);
             var pauseCommand = NewSyncplayCommand(SendCommandType.Pause);
-            SendCommand(session, BroadcastType.SingleSession, pauseCommand);
+            SendCommand(session, BroadcastType.CurrentSession, pauseCommand, CancellationToken.None);
```
Good question. I guess it depends on how locking happens underneath on the WebSocket. If I'm not wrong, one message at a time can be sent across the connection, so it's a matter of which message gets sent first (reception will match the same order).

I'll look into it, I was assuming this FIFO queue behaviour when calling `SendGroupUpdate` or `SendCommand`.

JustAMan replied:
> I was assuming this FIFO queue behaviour

I figured as much from the code, so I decided to raise a question earlier :) Finding root causes of bugs which have like 0.5% chance to happen is _hard_, so it's usually better to be extra-cautious in coding and reviewing :)

OancaAndrei replied:
The [docs](https://docs.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket.sendasync?view=netcore-3.1#System_Net_WebSockets_WebSocket_SendAsync_System_ArraySegment_System_Byte__System_Net_WebSockets_WebSocketMessageType_System_Boolean_System_Threading_CancellationToken_) say that: `Exactly one send and one receive is supported on each WebSocket object in parallel.`.

By looking at `.NET`'s [code](https://github.com/dotnet/corefx/blob/release/3.1/src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs) I found the `ManagedWebSocket` class that implements the `WebSocket` abstract class:
```
[ManagedWebSocket class]
Thread-safety:
    - It's acceptable to call ReceiveAsync and SendAsync in parallel.  One of each may run concurrently.
    - It's acceptable to have a pending ReceiveAsync while CloseOutputAsync or CloseAsync is called.
    - Attempting to invoke any other operations in parallel may corrupt the instance.  Attempting to invoke
      a send operation while another is in progress or a receive operation while another is in progress will
      result in an exception.
```
So it looks like calling `SendAsync` from _multiple_ threads could result in an exception being thrown.

In Jellyfin, a single `WebSocket` is wrapped and managed by the `SharpWebSocket` class, which has no locking mechanism implemented.

I'm surprised that everything _seems_ to work and no exception is thrown... :thinking: I guess that's because the WebSockets aren't used that often in other parts of the server.

`ManagedWebSocket` has some internal lock but I wouldn't rely on it, mainly because I don't see it mentioned in the docs. I guess it would be better for `SharpWebSocket` to implement a FIFO queue for outgoing messages.

Another solution would be to await for every single call to `SendGroupUpdate` or `SendCommand`, but that's ugly.

[This](https://www.codetinkerer.com/2018/06/05/aspnet-core-websockets.html) article looks helpful, mainly points 4, 5, 6 and 10. They talk about thread-safety, handling disconnections and Keepalives (apparently all the new code that handles them can be replaced by adding 3 lines in `Jellyfin.Serber/Startup.cs` when calling `app.UseWebSockets()` :sweat_smile:).

JustAMan replied:
> I'm surprised that everything _seems_ to work and no exception is thrown... 🤔 I guess that's because the WebSockets aren't used that often in other parts of the server.

We actually have some weird exceptions related to websockets logged time and again... right @Bond-009? So this might explain those.

dkanada commented:
Not sure if it was mentioned but if we keep SyncPlay as the name it should be capitalized with PascalCase, including on internal classes.

JustAMan commented:
> Not sure if it was mentioned but if we keep SyncPlay as the name it should be capitalized with PascalCase, including on internal classes.

@OancaAndrei this just means replace `Syncplay` with `SyncPlay` almost everywhere :)

OancaAndrei commented:
> Not sure if it was mentioned but if we keep SyncPlay as the name it should be capitalized with PascalCase, including on internal classes.

The idea was to pick a different name and I was waiting to complete the important stuff first.
I'll make sure to use the PascalCase for the naming.

dkanada commented:
I don't think we have a better name :laughing: the current name is very accurate and obvious for users so I think it's going to stick around.

JustAMan commented:
> I don't think we have a better name 😆

Yeah, we tried for a few days to invent a name and have thrown a few ideas off the wall, but it seems `SyncPlay` is the best choice right now :)

JustAPerson commented:
I run jellyfin on port 8096 and use a reverse proxy to access it from port 80/443. This feature was originally not working (GET `/socket` requests would 404) until I modified my reverse proxy settings. According to [this blog which helped me get it working](https://medium.com/@ibraheemabukaff/how-to-proxy-websockets-with-nginx-e333a5f0c0bb), websockets rely on some HTTP headers that are not forwarded through proxies by default.

Perhaps [the reverse proxy instructions](https://jellyfin.org/docs/general/networking/index.html) should be revised. The release notes should mention that websockets (which this relies on) may not work out of the box with an already working reverse proxy. 

I added the new `location /socket` part to my nginx config and now it works fine.
```
server {
  server_name MyServer.com;
  location /socket {
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_pass "http://localhost:8096/socket";
  }
  location / {
    proxy_pass http://localhost:8096/;
  }
}
```

OancaAndrei commented:
> I run jellyfin on port 8096 and use a reverse proxy to access it from port 80/443. This feature was originally not working (GET `/socket` requests would 404) until I modified my reverse proxy settings. According to [this blog which helped me get it working](https://medium.com/@ibraheemabukaff/how-to-proxy-websockets-with-nginx-e333a5f0c0bb), websockets rely on some HTTP headers that are not forwarded through proxies by default.
> 
> Perhaps [the reverse proxy instructions](https://jellyfin.org/docs/general/networking/index.html) should be revised. The release notes should mention that websockets (which this relies on) may not work out of the box with an already working reverse proxy.
> 
> I added the new `location /socket` part to my nginx config and now it works fine.
> 
> ```
> server {
>   server_name MyServer.com;
>   location /socket {
>     proxy_http_version 1.1;
>     proxy_set_header Upgrade $http_upgrade;
>     proxy_set_header Connection "upgrade";
>     proxy_pass "http://localhost:8096/socket";
>   }
>   location / {
>     proxy_pass http://localhost:8096/;
>   }
> }
> ```

I can confirm, my setup looks similar :smile:. I got the configuration directly from the Jellyfin [docs](https://jellyfin.org/docs/general/networking/nginx.html) tough.

JustAPerson commented:
I tested this with a total of 4 people watching tonight, and it works pretty well. Only problem I noticed was that if you advance to the next episode early, then that causes whoever pressed the button to leave the group.

So if you want to watch multiple episodes, it seems you may have to remake the group for each episode.

OancaAndrei commented:
> I tested this with a total of 4 people watching tonight, and it works pretty well.

I'm glad that it worked for you :grin:.

> Only problem I noticed was that if you advance to the next episode early, then that causes whoever pressed the button to leave the group.
>
> So if you want to watch multiple episodes, it seems you may have to remake the group for each episode.

This is indeed the main use case currently, for which films fit better. I know, it's kinda limited.

Playlist sync will surely be added, to ease out the experience, and should fix this issue with TV series as well.

JustAMan commented:
> Playlist sync will surely be added, to ease out the experience

But not in this PR, right?

JustAMan commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -505,7 +529,8 @@ private void HandleBufferingDoneRequest(SessionInfo session, PlaybackRequest req
             // Collected pings are used to account for network latency when unpausing playback
-            _group.UpdatePing(session, request.Ping ??= _group.DefaulPing);
+            var ping = request.Ping ?? _group.DefaulPing;
+            _group.UpdatePing(session, ping);
```
why did you want to do this? or is it Github showing something strange?

OancaAndrei replied:
You mean the extra `var`? I thought it had something to do with casting but the compiler is not complaining.
I can't really recall :thinking:. I'll make it a single line.

JustAMan commented on MediaBrowser.Api/Syncplay/SyncplayService.cs:
```diff
@@ -160,7 +154,7 @@ public class SyncplayService : BaseApiService
         {
             var currentSession = GetSession(_sessionContext);
-            _syncplayManager.NewGroup(currentSession);
+            _syncplayManager.NewGroup(currentSession, CancellationToken.None);
```
I wonder if `request` has some cancellation token... if it doesn't let's leave it how it is now though.

OancaAndrei replied:
So that's where that token should come from. `request` doesn't have one unfortunately.

JustAMan commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +140,205 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+            }
+            catch (WebSocketException exception)
+            {
+                _logger.LogWarning(exception, "Error sending ForceKeepAlive message to WebSocket {0}.", webSocket);
```
```suggestion
                _logger.LogWarning(exception, "Cannot send ForceKeepAlive message to WebSocket {0}.", webSocket);
```
logging a warning stating it was an error looks strange TBH

JustAMan commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +140,205 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+                    foreach (var webSocket in lost.ToList())
+                    {
+                        // TODO: handle session relative to the lost webSocket
+                        RemoveWebSocket(webSocket);
```
What if socket became "non-lost" after you had computed `lost` and `elapsed` but before we got here?
Maybe we should `lock (_webSocketsLock)` the whole function?

OancaAndrei replied:
Not knowing how long it takes for `SendForceKeepAlive` to complete, I've split the lock to avoid keeping onto it for too long, mainly because there's a `foreach` that awaits for all the sockets that need to be contacted.

Also, I had in mind this strict requirement: a client either keeps its time under `WebSocketLostTimeout` or it's considered to be lost.

So, with this rule in mind, using or not a lock after computing the `lost` list won't change the end result for those WebSockets. For the `elapsed` list what happens is that an extra `ForceKeepAlive` message will be sent, even when the client recently updated its status.

To be less restrictive, I could move the `lost = _webSockets.Where...` line in the second lock and merge it with the WebSockets lost from `inactive`.

Was this your concern?

OancaAndrei commented:
> > Playlist sync will surely be added, to ease out the experience
> 
> But not in this PR, right?

I was experimenting with it, but it definitely **won't be** in this PR, sorry if I made you worry.

JustAMan commented on MediaBrowser.Api/Syncplay/SyncplayService.cs:
```diff
@@ -207,7 +216,23 @@ public void Post(SyncplayLeaveGroup request)
+                }
+                catch (ArgumentNullException)
+                {
+                    Logger.LogError("ListGroups: null value for FilterItemId. Ignoring filter.");
```
is this an error or a warning? same about below

OancaAndrei replied:
Maybe a warning makes more sense, because the request is not rejected.

OancaAndrei commented:
> LGTM, minor only left.
> 
> Also please stop rebasing and stuffing all code review fixes in last commit, it forces me to re-read it in full every time instead of seeing what has changed :)

Snap, I'll keep this in mind for the future. I should've done it at the very end. Sorry for the burden.

JustAMan commented:
Well, actually, from my point of view you _can_ rebase things as long as you don't change commit patches (there's Github feature "marked as viewed" which helps here), but please don't incorporate multiple passes of feedback in single last commit, as instead of checking just response to my last feedback I'm forced to re-read the response to all feedbacks :D

OancaAndrei commented:
Last step is to replace `Syncplay` with `SyncPlay`.

Also, the issue with the WebSocket messages won't be tackled in this PR I guess.

ahti commented:
Wrt the name, something that's not `Syncplay` might be a better choice, to avoid confusion with [the other Syncplay](https://syncplay.pl).

JustAMan commented:
@ahti we had a several days long conversation internally and weren't able to come up with a name that most of the team liked... so we decided to stick to `SyncPlay`.

Bond-009 commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +140,205 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+            {
+                _logger.LogDebug("Watching {0} WebSockets.", _webSockets.Count);
+
+                inactive = _webSockets.Where(i =>
+                {
+                    var elapsed = (DateTime.UtcNow - i.LastKeepAliveDate).TotalSeconds;
+                    return (elapsed > WebSocketLostTimeout * ForceKeepAliveFactor) && (elapsed < WebSocketLostTimeout);
+                });
+                lost = _webSockets.Where(i => (DateTime.UtcNow - i.LastKeepAliveDate).TotalSeconds >= WebSocketLostTimeout);
```
This isn't thread safe `IEnumerable` is lazily evaluated. `.ToList()` these both inside the lock

Bond-009 commented on Emby.Server.Implementations/Syncplay/SyncplayController.cs:
```diff
@@ -0,0 +1,548 @@
+                        session => session.Session
+                    ).ToArray();
+                default:
+                    return new SessionInfo[] { };
```
Use `Array.Empty`

Bond-009 commented on Emby.Server.Implementations/Syncplay/SyncplayManager.cs:
```diff
@@ -0,0 +1,385 @@
+        private void OnSessionManagerSessionEnded(object sender, SessionEventArgs e)
+        {
+            var session = e.SessionInfo;
+            if (!IsSessionInGroup(session)) return;
```
Always use braces

Bond-009 commented on Emby.Server.Implementations/Syncplay/SyncplayManager.cs:
```diff
@@ -0,0 +1,385 @@
+        private void OnSessionManagerPlaybackStopped(object sender, PlaybackStopEventArgs e)
+        {
+            var session = e.Session;
+            if (!IsSessionInGroup(session)) return;
```
Same

Bond-009 commented on Emby.Server.Implementations/Syncplay/SyncplayManager.cs:
```diff
@@ -0,0 +1,385 @@
+
+                if (IsSessionInGroup(session))
+                {
+                    if (GetSessionGroup(session).Equals(groupId)) return;
```
Same

Bond-009 commented on MediaBrowser.Api/Syncplay/SyncplayService.cs:
```diff
@@ -0,0 +1,312 @@
+using MediaBrowser.Model.Syncplay;
+using Microsoft.Extensions.Logging;
+
+namespace MediaBrowser.Api.Syncplay
```
It's too late now, but new APIs should use ASP.net instead of the old framework

Bond-009 commented on MediaBrowser.Api/Syncplay/SyncplayService.cs:
```diff
@@ -0,0 +1,312 @@
+                {
+                    Logger.LogWarning("ListGroups: null value for FilterItemId. Ignoring filter.");
+                }
+                catch (FormatException)
```
Use TryParse instead of catching exceptions

Bond-009 commented on MediaBrowser.Api/Syncplay/SyncplayService.cs:
```diff
@@ -0,0 +1,312 @@
+                    Logger.LogError("JoinGroup: null value for PlayingItemId. Ignoring request.");
+                    return;
+                }
+                catch (FormatException)
```
Use `TryParse` instead of catching exceptions

Bond-009 commented on MediaBrowser.Controller/Net/IWebSocketConnection.cs:
```diff
@@ -26,6 +26,12 @@ public interface IWebSocketConnection : IDisposable
+        /// Gets or sets the date of last Keeplive received.
+        /// </summary>
+        /// <value>The date of last Keeplive received.</value>
+        public DateTime LastKeepAliveDate { get; set; }
```
```suggestion
        DateTime LastKeepAliveDate { get; set; }
```
Too stay consistent with other interface members

Bond-009 commented on MediaBrowser.Controller/Syncplay/GroupInfo.cs:
```diff
@@ -0,0 +1,150 @@
+        /// <param name="session">The session.</param>
+        public void AddSession(SessionInfo session)
+        {
+            if (ContainsSession(session.Id.ToString())) return;
```
braces

Bond-009 commented on MediaBrowser.Controller/Syncplay/GroupInfo.cs:
```diff
@@ -0,0 +1,150 @@
+
+        public void RemoveSession(SessionInfo session)
+        {
+            if (!ContainsSession(session.Id.ToString())) return;
```
same

Bond-009 commented on MediaBrowser.Controller/Syncplay/GroupInfo.cs:
```diff
@@ -0,0 +1,150 @@
+        {
+            if (!ContainsSession(session.Id.ToString())) return;
+            GroupMember member;
+            Participants.Remove(session.Id.ToString(), out member);
```
```suggestion
            Participants.Remove(session.Id.ToString(), out _);
```

Bond-009 commented on MediaBrowser.Controller/Syncplay/GroupInfo.cs:
```diff
@@ -0,0 +1,150 @@
+        /// <param name="ping">The ping.</param>
+        public void UpdatePing(SessionInfo session, long ping)
+        {
+            if (!ContainsSession(session.Id.ToString())) return;
```
Braces

Bond-009 commented on MediaBrowser.Controller/Syncplay/GroupInfo.cs:
```diff
@@ -0,0 +1,150 @@
+        /// <param name="isBuffering">The state.</param>
+        public void SetBuffering(SessionInfo session, bool isBuffering)
+        {
+            if (!ContainsSession(session.Id.ToString())) return;
```
braces

Bond-009 commented on MediaBrowser.Controller/Syncplay/GroupInfo.cs:
```diff
@@ -0,0 +1,150 @@
+        {
+            foreach (var session in Participants.Values)
+            {
+                if (session.IsBuffering) return true;
```
braces

Bond-009 commented on MediaBrowser.Controller/Syncplay/ISyncplayManager.cs:
```diff
@@ -0,0 +1,69 @@
+        /// <param name="groupId">The group id.</param>
+        /// <param name="request">The request.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        void JoinGroup(SessionInfo session, string groupId, JoinGroupRequest request, CancellationToken cancellationToken);
```
Why is `groupId` a string instead of a `Guid`? (Not just here, but almost everywhere)

OancaAndrei replied:
Good point. I'll update these.

Bond-009 commented on MediaBrowser.Model/Syncplay/GroupInfoView.cs:
```diff
@@ -0,0 +1,38 @@
+        /// Gets or sets the participants.
+        /// </summary>
+        /// <value>The participants.</value>
+        public string[] Participants { get; set; }
```
Use a `ReadOnlyList` instead of an array for properties

dkanada commented on Emby.Server.Implementations/HttpServer/WebSocketConnection.cs:
```diff
@@ -174,7 +172,14 @@ private void OnReceiveInternal(string message)
                 };
 
-                OnReceive(info);
+                if (info.MessageType.Equals("KeepAlive", StringComparison.Ordinal))
```
I think strict checks are a good idea. It doesn't help to be vague with the API functionality, we should clearly state exactly what works and what doesn't work.

OancaAndrei replied:
BTW, I'm afraid that all this `KeepAlive` part is going to be removed by using the same functionality that the protocol already provides.

I haven't tested it yet, but adding the following to `Jellyfin.Server/Startup.cs` should handle things without the need of this watcher or additional client code either.
```
app.UseWebSockets(new WebSocketOptions
{
   KeepAliveInterval = TimeSpan.FromSeconds(20),
});
```

The 20 seconds can be read from some configuration somewhere.

Should I test this, to avoid clutter code now, or leave the change for another PR?

JustAMan replied:
This code is needed because we want to handle dead websockets here.
Also docs say that default value of keepalive is 2 minutes, but I don't think it really works...

So let's leave current code for now.

OancaAndrei replied:
Agreed.

OancaAndrei commented:
I cannot believe the timing, I've rebased again without noticing these requested changes. I swear I'm not doing this on purpose :sweat_smile:.

OancaAndrei commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +140,205 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+        /// Adds a WebSocket to the KeepAlive watchlist.
+        /// </summary>
+        /// <param name="webSocket">The WebSocket to monitor.</param>
+        private void KeepAliveWebSocket(IWebSocketConnection webSocket)
```
It was, but it got changed, amongst other reasons, because by `.Wait()`-ing there we can catch a possible exception, instead of letting it slip with the async task.

Another reason is that Sonar complained about not awaiting for the async task when calling `KeepAliveWebSocket` from `OnServerManagerWebSocketConnected` which is not `async` as well. And this issue was seen as a `Major` code smell.
```
Because this call is not awaited, execution of the current method continues before the call is completed.
Consider applying the 'await' operator to the result of the call.
```

Bond-009 replied:
`OnServerManagerWebSocketConnected` can be made async

OancaAndrei replied:
This solves my dilemma. I'll make it async then.

JustAMan commented on MediaBrowser.Api/SyncPlay/SyncPlayService.cs:
```diff
@@ -128,185 +128,184 @@ public class SyncplayUpdatePing : IReturnVoid
+            Guid groupId;
+            Guid playingItemId = Guid.Empty;
+
+            var valid = Guid.TryParse(request.GroupId, out groupId);
+            if (!valid)
```
```suggestion
            if (!Guid.TryParse(request.GroupId, out groupId))
```

JustAMan commented on MediaBrowser.Api/SyncPlay/SyncPlayService.cs:
```diff
@@ -128,185 +128,184 @@ public class SyncplayUpdatePing : IReturnVoid
-                    return;
-                }
-                catch (FormatException)
+                valid = Guid.TryParse(request.PlayingItemId, out playingItemId);
+                if (!valid)
```
```suggestion
                if (!Guid.TryParse(request.PlayingItemId, out playingItemId))
```

JustAMan commented on MediaBrowser.Api/SyncPlay/SyncPlayService.cs:
```diff
@@ -128,185 +128,184 @@ public class SyncplayUpdatePing : IReturnVoid
-                    Logger.LogWarning("ListGroups: null value for FilterItemId. Ignoring filter.");
-                }
-                catch (FormatException)
+                var valid = Guid.TryParse(request.FilterItemId, out filterItemId);
+                if (!valid)
```
```suggestion
                if (!Guid.TryParse(request.FilterItemId, out filterItemId))
```

sparky8251 commented:
By no means interpret what I'm about to say as a blocker let alone a request. Just throwing out an idea to hopefully solve the bandwidth problem of group playback (CPU/GPU problem is solved with reproducible chunking).

Would client side p2p (likely via bittorrent) help with playback sync and distributing the bandwidth load for large groups? We have a web server that's exposed to all the clients already, so setting up a tracker and making sure its universally accessible is pretty easy.

A rudimentary non-bittorrent p2p might also make it easier to sync clients and server up depending on geographical layout of the members.

xnaas commented:
As long as P2P was an option and not a forced requirement. There are plenty of networks out there that block all P2P traffic entirely (many college campuses, for instance).

JustAMan commented:
> Would client side p2p (likely via bittorrent) help with playback sync and distributing the bandwidth load for large groups?

It would, but
* can not be a hard requirement (some networks block P2P)
* should be done in a separate PR
* has limited use because different clients could be utilizing different streams of same media file (e.g. different versions or different transcode codecs or quality)

dkanada commented:
I don't think that should be an option. We shouldn't try to implement P2P functionality in clients for a single feature, even if it would be somewhat useful.

MayeulC commented:
> By no means interpret what I'm about to say as a blocker let alone a request. Just throwing out an idea to hopefully solve the bandwidth problem of group playback (CPU/GPU problem is solved with reproducible chunking).
> 
> Would client side p2p (likely via bittorrent) help with playback sync and distributing the bandwidth load for large groups? We have a web server that's exposed to all the clients already, so setting up a tracker and making sure its universally accessible is pretty easy.
> 
> A rudimentary non-bittorrent p2p might also make it easier to sync clients and server up depending on geographical layout of the members.

That's [peertube](https://joinpeertube.org/) you're talking about. It works quite well, I'd be thrilled if some code could be shared (maybe as a plugin). Alternatively, something like [hypercore](https://hypercore-protocol.org/) could also be leveraged, but that might be less of a good fit :)

Nautigsam commented:
SyncPlay and peer to peer are different features. It would be a shame to delay the merge of a nearly-completed feature because of a late scope extension.

dkanada commented:
> It would be a shame to delay the merge of a nearly-completed feature because of a late scope extension.

It was never getting delayed, we're just talking about future plans here. This will most definitely be in the next major release.

Bond-009 commented on Emby.Server.Implementations/HttpServer/WebSocketConnection.cs:
```diff
@@ -218,7 +221,42 @@ private async Task ProcessInternal(PipeReader reader)
+        private void SendKeepAliveResponse()
+        {
+            LastKeepAliveDate = DateTime.UtcNow;
+            SendAsync(new WebSocketMessage<string>
```
Doesn't this need to be awaited?

OancaAndrei replied:
I wonder if there was a reason, but it can surely be awaited.

Bond-009 commented on Emby.Server.Implementations/Session/SessionWebSocketListener.cs:
```diff
@@ -99,5 +141,206 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+        /// <param name="e">The event arguments.</param>
+        private void OnWebSocketClosed(object sender, EventArgs e)
+        {
+            var webSocket = (IWebSocketConnection) sender;
```
```suggestion
            var webSocket = (IWebSocketConnection)sender;
```

Bond-009 commented on Emby.Server.Implementations/SyncPlay/SyncPlayController.cs:
```diff
@@ -0,0 +1,548 @@
+        /// Releases unmanaged and optionally managed resources.
+        /// </summary>
+        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+        protected virtual void Dispose(bool disposing)
```
Why is this an `IDisposable` if it disposes nothing?

OancaAndrei replied:
I thought that this was the C# way to "deconstruct" objects and I usually just write it, in case it's needed along the way (also, I copy-pasted from some other existing class at the beginning). I'll clean it.

Bond-009 commented on Emby.Server.Implementations/SyncPlay/SyncPlayController.cs:
```diff
@@ -0,0 +1,548 @@
+            {
+                var runTimeTicks = _group.PlayingItem.RunTimeTicks ?? 0;
+                ticks = ticks > runTimeTicks ? runTimeTicks : ticks;
+            }
```
```suggestion
            }

```

Bond-009 commented on Emby.Server.Implementations/SyncPlay/SyncPlayController.cs:
```diff
@@ -0,0 +1,548 @@
+                PlayingItemName = _group.PlayingItem.Name,
+                PlayingItemId = _group.PlayingItem.Id.ToString(),
+                PositionTicks = _group.PositionTicks,
+                Participants = _group.Participants.Values.Select(session => session.Session.UserName).Distinct().ToList().AsReadOnly()
```
```suggestion
                Participants = _group.Participants.Values.Select(session => session.Session.UserName).Distinct().ToList()
```
Everywhere in the code we trust that nothing will try to modify ReadOnly collections, `AsReadOnly` adds another abstraction which slows down all access to the inner collection. Just to stay consistent

Bond-009 commented on Emby.Server.Implementations/SyncPlay/SyncPlayManager.cs:
```diff
@@ -0,0 +1,397 @@
+            if (IsSessionInGroup(session))
+            {
+                throw new InvalidOperationException("Session in other group already!");
+            }
```
```suggestion
            }

```

Bond-009 commented on MediaBrowser.Api/SyncPlay/SyncPlayService.cs:
```diff
@@ -0,0 +1,302 @@
+            }
+
+            // Both null and empty strings mean that client isn't playing anything
+            if (!String.IsNullOrEmpty(request.PlayingItemId) && !Guid.TryParse(request.PlayingItemId, out playingItemId))
```
```suggestion
            if (!string.IsNullOrEmpty(request.PlayingItemId) && !Guid.TryParse(request.PlayingItemId, out playingItemId))
```

Bond-009 commented on MediaBrowser.Api/SyncPlay/SyncPlayService.cs:
```diff
@@ -0,0 +1,302 @@
+            var currentSession = GetSession(_sessionContext);
+            var filterItemId = Guid.Empty;
+
+            if (!String.IsNullOrEmpty(request.FilterItemId) && !Guid.TryParse(request.FilterItemId, out filterItemId))
```
```suggestion
            if (!string.IsNullOrEmpty(request.FilterItemId) && !Guid.TryParse(request.FilterItemId, out filterItemId))
```

Bond-009 commented on MediaBrowser.Controller/SyncPlay/GroupInfo.cs:
```diff
@@ -0,0 +1,168 @@
+        /// <summary>
+        /// Default ping value used for sessions.
+        /// </summary>
+        public readonly long DefaulPing = 500;
+        /// <summary>
+        /// Gets or sets the group identifier.
+        /// </summary>
+        /// <value>The group identifier.</value>
+        public readonly Guid GroupId = Guid.NewGuid();
```
Fields should not be public

OancaAndrei replied:
Just found out that auto-properties exist, exactly what I needed.

Bond-009 commented on MediaBrowser.Controller/SyncPlay/GroupInfo.cs:
```diff
@@ -0,0 +1,168 @@
+                {
+                    return true;
+                }
+            }
```
```suggestion
            }

```

Aterfax commented:
I am happy to test before release (so long as you can get me a docker) - been looking forward to this for a long while.

Vitorvlv commented:
I believe `docker pull artiume/jellyfin:syncplay` should give you jellyfin with this feature.

Aterfax commented:
> 
> 
> I believe `docker pull artiume/jellyfin:syncplay` should give you jellyfin with this feature.

Cheers, I will work out how to pull that and get it working with my existing install on unraid then feedback my experience.

reavessm commented:
idk where to post issues on this pull request, but I tried testing on artiume/jellyfin:syncplay docker and I'm having some trouble.  I made a new user, started playing something, started a group (which I didn't know you had to be playing something to create it but you definitely did say that in the initial message so that's on me I guess), then when I tried to join that group as the other user I just got a spinning wheel.  I am using a reverse-proxy and I just opened the group from the home menu without opening the file

Aterfax commented:
Got this installed now - start a sync play group - how do I get another user to join?

Second user seems to be stuck at a loading wheel after hitting the button?

dkanada commented:
Could be a cache issue, try opening the developer tools on your browser and navigating to the network tab, then selecting disable cache at the top.

Aterfax commented:
> 
> 
> Could be a cache issue, try opening the developer tools on your browser and navigating to the network tab, then selecting disable cache at the top.

Tested without cache - not that.

As with the error message above, seems to be server side?

``` [05:36:21] [ERR] [9] Emby.Server.Implementations.HttpServer.HttpListenerHost: Could not find handler for /Syncplay/htmlvideoplayer/GetUtcTime
[05:36:21] [ERR] [9] Emby.Server.Implementations.HttpServer.HttpListenerHost: Error processing request: Unable to find the specified file.
 
```

dkanada commented:
We haven't renamed those classes yet for the same reason Emby never renamed half the projects from MediaBrowser to their new name. It would take a huge amount of effort to refactor, review, merge, and then fix all existing pull requests with the new name, whereas leaving them alone doesn't technically harm anything for now. Hopefully at some point in the future we can find the time though.

Xyz00777 commented:
i haven't read everything carefully but if this p2p feature should be integrated (the part of the stream to be shared with others via the users) for watching at the same time then i have the question if the p2p is also encrypted???, otherwise you might get a letter or a visit from the police. here it would be great if it is also encrypted by default, if not you should be able to disable this on the server so that the stream is only sent from the server (preferably with a note in the setting that this is not encrypted when p2p is enabled)

mcarlton00 commented:
The p2p is an expansion to this far down the line.  If we decide to go that route it will definitely need it's own discussion and considerations (encryption is one of those).

Aterfax commented:
> 
> 
> i haven't read everything carefully but if this p2p feature should be integrated (the part of the stream to be shared with others via the users) for watching at the same time then i have the question if the p2p is also encrypted???, otherwise you might get a letter or a visit from the police. here it would be great if it is also encrypted by default, if not you should be able to disable this on the server so that the stream is only sent from the server (preferably with a note in the setting that this is not encrypted when p2p is enabled)

 Please do not fearmonger needlessly!

You will not get a visit from the police or a letter from your ISP, copyright theft claims are made by companies polling the bit torrent tracker lists for IP addresses associated with torrent files which contain stolen content. 

The chances of you getting a letter or visit from the police is effectively nil in a home media server implementation (you are sharing your media with authorized users of your media centre only.) 

This is also basically undetectable anyway unless you are living in a dystopic nation state that is doing deep packet inspection, has a lot of spare money and a lot of spare time to chase people for something that is most likely going to be determined as legal anyway (sharing with family.)

It should however be encrypted anyway and this is supported by bittorrent protocols.

Implementations of web based torrenting already exist with projects like peer tube or bittorrent web which may be of use in implementing such a system in jellyfin.

Related to my error message above, anyone got any ideas?

SaddFox commented:
> 
> 
> > Could be a cache issue, try opening the developer tools on your browser and navigating to the network tab, then selecting disable cache at the top.
> 
> Tested without cache - not that.
> 
> As with the error message above, seems to be server side?
> 
> Error message in the jellyfin logs - incidentally, emby might need purging ;)
> 
> ```
> [05:36:21] [ERR] [9] Emby.Server.Implementations.HttpServer.HttpListenerHost: Could not find handler for /Syncplay/htmlvideoplayer/GetUtcTime
> [05:36:21] [ERR] [9] Emby.Server.Implementations.HttpServer.HttpListenerHost: Error processing request: Unable to find the specified file.
>  
> ```

I had the same issue when I gave it a try. Btw that docker image is a month old now.

Aterfax commented:
> 
> 
> > > Could be a cache issue, try opening the developer tools on your browser and navigating to the network tab, then selecting disable cache at the top.
> > 
> > 
> > Tested without cache - not that.
> > As with the error message above, seems to be server side?
> > Error message in the jellyfin logs - incidentally, emby might need purging ;)
> > ```
> > [05:36:21] [ERR] [9] Emby.Server.Implementations.HttpServer.HttpListenerHost: Could not find handler for /Syncplay/htmlvideoplayer/GetUtcTime
> > [05:36:21] [ERR] [9] Emby.Server.Implementations.HttpServer.HttpListenerHost: Error processing request: Unable to find the specified file.
> >  
> > ```
> 
> I had the same issue when I gave it a try. Btw that docker image is a month old now.

Hmm, @OancaAndrei do you have a docker build for testing handy?

xnaas commented:
I forked https://github.com/OancaAndrei/jellyfin/tree/syncplay and set it to build @ https://hub.docker.com/r/xnaas/jellyfin-syncplay @Aterfax.

OancaAndrei commented:
> > Could be a cache issue, try opening the developer tools on your browser and navigating to the network tab, then selecting disable cache at the top.
> 
> Tested without cache - not that.
> 
> As with the error message above, seems to be server side?
> 
> Error message in the jellyfin logs - incidentally, emby might need purging ;)
> 
> ```
> [05:36:21] [ERR] [9] Emby.Server.Implementations.HttpServer.HttpListenerHost: Could not find handler for /Syncplay/htmlvideoplayer/GetUtcTime
> [05:36:21] [ERR] [9] Emby.Server.Implementations.HttpServer.HttpListenerHost: Error processing request: Unable to find the specified file.
>  
> ```

Sorry for the late reply. That build looks indeed old, my guess is that the web client hasn't been updated. I'll have a look at how to setup a working one. Hang in there :grin:.

dkanada commented:
Lets not turn this pull request into a general troubleshooting issue guys, we can tackle issues easier once this entire feature gets merged into the master branch.

JustAMan commented:
@OancaAndrei please address remaining comments from @Bond-009 so we can finally merge, thanks! :)

Aterfax commented:
> > > Could be a cache issue, try opening the developer tools on your browser and navigating to the network tab, then selecting disable cache at the top.
> > 
> > 
> > Tested without cache - not that.
> > As with the error message above, seems to be server side?
> > Error message in the jellyfin logs - incidentally, emby might need purging ;)
> > ```
> > [05:36:21] [ERR] [9] Emby.Server.Implementations.HttpServer.HttpListenerHost: Could not find handler for /Syncplay/htmlvideoplayer/GetUtcTime
> > [05:36:21] [ERR] [9] Emby.Server.Implementations.HttpServer.HttpListenerHost: Error processing request: Unable to find the specified file.
> >  
> > ```
> 
> Sorry for the late reply. That build looks indeed old, my guess is that the web client hasn't been updated. I'll have a look at how to setup a working one. Hang in there 😁.
> 
> ~EDIT: uploaded to `andreioanca/jellyfin:syncplay` but I noticed some issues with the initial setup, which I'm not sure how to fix. The issue seems to be this one #3132, I'll merge from `master` and report back.~
> EDIT2: @Aterfax, @reavessm, @SaddFox I've uploaded a build to `andreioanca/jellyfin:syncplay`:
> 
> ```
> docker pull andreioanca/jellyfin:syncplay
> ```

Thanks, I'll look at loading it up shortly for testing.

JustAPerson commented:
@OancaAndrei thank you so much for your monumental effort and perseverance! I love this feature so  much already :heart:

StevenGreenbank commented:
I have a question regarding this feature: in the scenario where one person, a remote user, in the watch group has a downloaded, local copy of the video (similar to plex's sync feature) and the other user is watching the video off of their home network, does the jellyfin app play the remote user's local copy? 

I'm making an assumption here that jellyfin has a sync feature akin to plex's, of course. But if the jellyfin app DOES indeed play the remote user's local, synced copy I would be compelled to use this over plex.

itsboo07 commented:
@OancaAndrei and @joshuaboniface firstly thanks for this feature and it works great with web player. But web player lacks few options and features. Is it possible to have sync feature with external players like vlc/mx player?

OancaAndrei commented:
Hi @itsboo07, have you considered [Jellyfin Media Player](https://jellyfin.org/posts/client-jmp/)?
Regarding VLC or MX Player, I've never looked into their plugin system. The first roadblock would be session handling and WebSockets as SyncPlay uses them to sync the devices.

itsboo07 commented:
@OancaAndrei thanks for responding... I did not try jellyfin media player as I use android device to watch videos or movies or series. And jellyfin media player isn't available for android(as of now I guess) . I tried integrated media player from client settings in android and sync Playback doesn't work their either.... Only in simple web player it works... But the web player has big subtitle fonts and some functions lacking.
